### ISSUE TITLE
Stable memory for event log.

### ISSUE THREAD

**Issue by matthewhammer** [matthewhammer/candid-spaces#6]
The event log permits re-generation of each other table/relation/map stored in the canister state.

Providing stable memory for the event log means that all events are stored in stable memory, and that upgrades need not transform it.

Consequently, a canister with this feature is a canister with a lossless representation of its past state.  That has important correctness implications for upgrades, since all other information can be regenerated from it, should the other information (stored in auxiliary data structures that may change format) be damaged or lost in a faulty upgrade.

**Comment by matthewhammer** [comment]
This issue is tied to a larger idea, and a larger motivation, about inserting established incremental computation techniques into the prototypical canister design and its lifecycle of upgrades.

The benefit is a simple design that is efficient and correct, but also flexible and supportive for canister feature evolution.

In particular, there is a tension between adding features, and the complexity of doing upgrades and managing state in stable memory.  In general, it would be simplest if adding new features only meant _minimal_ extensions to the data types within stable memory.

This issue proposes that we store the event log in stable memory, so as to always have a source of truth from which to upgrade and recover, later.  So for instance, a minimal addition would be adding a new log record and variant type for each new major operation performed by the canister, so as to record it, and re-produce its effects upon a replay (during a migration or upgrade, etc).

#### Please Repeat Yourself (PRY) 

Here's a novel (?) idea about how to simplify and systematize the design of stable memory for canister state that undergoes lots of:
 - feature evolution (adding more canister API methods, and state to support them)
 - regular usage by real people that want the service to be stable and reliable.

Suppose that the event log is the only developer-defined type stored in stable memory.  That way, adding new features is a minimal change to that type (see above: add a new record type for the operation, and a variant to include it within the type of canister events).

If we additionally store a [(library-defined) incremental computing cache](https://github.com/matthewhammer/motoko-adapton) in stable memory as a companion to this event log, and we use that incremental cache in the upgrade hooks (reading from it to avoid re-processing the log), we should (in theory) be able to make "automatically incremental" the process of walking over the entire event log from the start of the canister's life, and re-generating the acceleration structures that provide efficient access to canister state (the tables/maps used by update and query calls, post upgrade).

Thoughts @chenyan-dfinity ?

**Comment by matthewhammer** [comment]
Regarding the "please repeat yourself" idea above (what I would call the idea of complete replay as the "ordinary semantics" of incremental updates) --- a former student colleague and I thought about this problem together, and it seemed promising to us as the time.

Here are some [initial results](http://kyleheadley.github.io/PHDWebsite/PRY-submit.pdf), along with unrelated ideas, circa 2016.  For a variety of reasons, that work did not continue.  But I think the ideas remain very promising, and always have.

**Comment by chenyan-dfinity** [comment]
I don't quite see how this works. Let's start with an example. In version 1, my canister uses an array of key-value pairs to store the user profiles. In version 2, I upgrade the profile to use `HashMap`. I think the event-log in the stable memory is unchanged. How do we propagate the event log to the concrete data structure?

**Comment by matthewhammer** [comment]
> In version 1, my canister uses an array of key-value pairs to store the user profiles. In version 2, I upgrade the profile to use HashMap.

I propose a different example.

In version 1, we have an app with users and profiles, but no follows relation.

In version 2, we add a follows relation, and the API for updating it.  Those operations become new events in the event log type (the only thing in stable memory other than the adapton cache).  Those operations have an interpretation (like operations in a DSL, which they are) and their interpretation is to populate the HashMaps and other structures that represent the follows relation.

During an upgrade from version 2 into version 3, where the follows relation does not change definition, we recover the cached representation from its initial construction, as a cached interpretation of the same log event.

Summary:
- in transition from version 1 to version 2, we update with a new feature (follows relation); there is no incremental reuse yet (if there is, let's ignore it for now).  But the update is simple because we need not add that relation directly to the representation of stable memory, or the upgrade hooks.
- in transition from version 2 to version 3, we get to recover the follows relation representation from the cache, and it's important that to get further reuse, we do not change this representation's type in subsequent upgrades; it must stay representing the same thing, in the same way.  That's fine.  What's more likely is adding additional relations, which would be similar to the original transition from version 1 to version 2.

I should also add that incremental reuse during upgrades is only one time when these caches would help.  In general, they'd help for hypothetical transactions that get rolled back and replayed, for similar reasons.

**Comment by chenyan-dfinity** [comment]
* I can see how this works for schema changes. But if we want to refactor the code to a different data structure, like in my example, it's unclear how to do it.
* As you mentioned, the upgrade is a one-time job. Do we really care about incremental reuse? The adapton cache size is probably more than the event-log size, because of meta data and closures? 
* The real difficulty in upgrade is that we lose all correctness guarantees provided by the type system. If we mess up during an upgrade, there is no way to rollback. If the event log idea can somehow give a correctness guarantee, that would be a big selling point. I think the efficiency of data migration is not a concern here. It's fine to take a few days to migrate the data, as long as the data is not lost.

**Comment by chenyan-dfinity** [comment]
I think the whole concept of upgrade is wrong. It's imperative with no persistency. The more functional way of "upgrade" would be to start a new canister, and importing data from the old canister. After we validate the data is intact, we tell the system to switch routing to the new canister.

**Comment by matthewhammer** [comment]
> But if we want to refactor the code to a different data structure, like in my example, it's unclear how to do it.

I agree.  That kind of change seems to be common in practice, but perhaps with some incentive to migrate data structure designs more _incrementally_, people would choose those more incremental design paths.  Only time and more experience will tell.  But I totally understand that there are serious limitations to what I am proposing.  This is one of them, certainly.

>  the upgrade is a one-time job. Do we really care about incremental reuse?

Good question.  Maybe not, but maybe so, depending on the developer and how frequently they push new Wasm.

Imagine a live programming system that can do incremental Wasm builds.  In that case, an ordinary session of development could be to push Wasm binaries quite frequently, and rely on upgrades being incremental and most data being stable (all bulk long-term data in stable memory, and never moved out of it) for efficiency and responsiveness in the face of that high frequency.

That's the direction I want to push for "live programming the IC", at least eventually, if not the medium term.  Perhaps other live experiences are possible too (with no Wasm recompilation, for instance, by using an interpreter for a live programming DSL that stays fixed as the DSL AST changes over time); but I find myself wanting to push new Motoko-based Wasm quite frequently as I play with canisters that I create in Motoko, either "locally" or remotely, on the main network.

It would be really cool to have the model where reinstall + upgrade is nearly "real time" (a second or a few, not minutes).

**Comment by matthewhammer** [comment]
> I think the whole concept of upgrade is wrong. It's imperative with no persistency. The more functional way of "upgrade" would be to start a new canister, and importing data from the old canister. After we validate the data is intact, we tell the system to switch routing to the new canister.

I see your complaint.

I think the salient difference is how the canister identity changes or not.  If it does (as you propose), then that's a functional update versus an imperative one (where the canister identity stays fixed, but its "meaning" changes).

Surely you can support functional updates as a "mode of use" of the model we have now?  You need to somehow migrate the state, but that could be done on demand, perhaps?

**Comment by matthewhammer** [comment]
> After we validate the data is intact, we tell the system to switch routing to the new canister.

What would that entail, generally?  Perhaps a simple example would help me see what you envision there.

**Comment by matthewhammer** [comment]
> If the event log idea can somehow give a correctness guarantee, that would be a big selling point. I think the efficiency of data migration is not a concern here. It's fine to take a few days to migrate the data, as long as the data is not lost.

As usual, you are the better salesperson.

Now that you say it, I agree -- focusing on the promise of "lossless representation of canister history" is the key selling point, as you suggest.

I updated the Issue description to reflect the long-term correctness motivation as a primary one.

**Comment by matthewhammer** [comment]
[This PR](https://github.com/matthewhammer/motoko-sequence/pull/24) shows that the sequence type works as a `stable var`, though that fact was never in doubt.  Just a sanity test, and initial baby step to try the idea somewhere.

**Comment by chenyan-dfinity** [comment]
> It would be really cool to have the model where reinstall + upgrade is nearly "real time" (a second or a few, not minutes).

This seems to be a different problem though. Update call (Installing Wasm) takes a few seconds, even if the there is no stable memory to migrate. It's hard for IC to be "real time".

> functional upgrade

My complaint is that nobody is doing in-place upgrade in today's world. It doesn't work. You cannot rollback if there is a bug. If you deploy a new canister, you can throughly test the new canister before using the new canister in production.

Instead of pre-/post-upgrade hook, each canister exposes `import` and `export` methods, so that it can migrate data from one canister to the other.

**Comment by chenyan-dfinity** [comment]
> After we validate the data is intact, we tell the system to switch routing to the new canister.
> What would that entail, generally? Perhaps a simple example would help me see what you envision there.

Imagine a proxy service that simply relay request to the old canister. When you deploy a new/upgraded canister and migrate data from the old canister, you can ask the proxy to relay request to both canisters, so that we can monitor if any runtime metric look good, or any diff between the old and new canisters. After you have enough confident you can stop relaying requests to the old canister.

This runtime validation is never possible with in-place upgrade. Plus there is no static correctness guarantee, why would anyone want to upgrade canisters in-place?

**Comment by matthewhammer** [comment]
> why would anyone want to upgrade canisters in-place?

1. A developer that is just getting going, with no users.
2. Canister IDs are system resources that cost something to create?  Not sure about this actually.

I think this is a very good question, FWIW.  These are a couple of answers that come to my mind, but I think you raise some very good points about another approach (the `import` and `export` idea).

**Comment by matthewhammer** [comment]
> Imagine a proxy service ...

Yes, I see the value in such a service.  I expect that this will become a business area on the IC, not just something that we need to "solve" once and for all.

We are not replacing Amazon web services with these upgrade hooks, some other businesses will emerge to use those hooks as very basic primitives to provide developers with an experience that is more "managed", like the one in your proxy example (and more like Amazon/Google, or so I imagine).

**Comment by matthewhammer** [comment]
Interestingly, I have upgraded [my instance of CandidSpaces](https://ic.rocks/principal/fzcsx-6yaaa-aaaae-aaama-cai) a few times now, and once with a `stable var` type change that may not be compatible (going from a `[ X ]` to a `record { fieldName : [ X ]}`?), but it seems that no problems have occurred yet.  The log still seems readable, as I [`put` entire trees of files into it now](https://github.com/matthewhammer/candid-spaces/pull/14).

**Comment by matthewhammer** [comment]
Spoke too soon.  I lost the `eventLog` while doing an upgrade tonight.  I'm not entirely sure why, since I think I was changing other things, and not the `stable var`s.
